### ISSUE TITLE
Support redis cluster

### ISSUE THREAD

**Issue by zfanta** [kkoomen/nestjs-throttler-storage-redis#660]
When `ThrottlerStorageRedisService` get an instance of `Redis.Cluster`, the instance meets the last `else`, then it tries connect default host(127.0.0.0) and port(6379)

https://github.com/kkoomen/nestjs-throttler-storage-redis/blob/80aef95ae13200b813829fa8163b62d17b9f868c/src/throttler-storage-redis.service.ts#L14-L23

**Comment by kkoomen** [comment]
@zfanta I've never met the use case to use the `Redis.Cluster`, nor have I experience in that. Would you say it's basically adjusting the if-statement like so?

```ts
if (redisOrOptions instanceof Redis || redisOrOptions instanceof Redis.Cluster) { 
  this.redis = redisOrOptions; 
}
```
```

**Comment by mikeulvila** [comment]
For anyone else that comes across this looking to use with a cluster, `SCAN` command is not supported for a cluster as it iterates the set of keys for a single Redis database.

**Comment by uncodable** [comment]
So does this storage package support `Redis.Cluster`? @kkoomen @mikeulvila thanks in advance

**Comment by kkoomen** [comment]
@mikeulvila so how would this be implemented? I'm fine if people want support for clusters, but I have no clue about how to implement it as I have no experience with it.

**Comment by uncodable** [comment]
@kkoomen ok so I got this working, when you use a cluster keyPrefix `{login-throttle}` using `Redis.Cluster` however I ended up making my own lib for this. Will leave this here for anyone who needs this in the future:

```ts
  public constructor(nodes: ClusterNode[], options?: ClusterOptions, scanCount?: number) {
    this.scanCount = typeof scanCount === 'undefined' ? 1000 : scanCount;
    this.redis = new Redis.Cluster(nodes, options);
    this.disconnectRequired = true;
  }
 ```

```ts
return {
        ttl: rateLimitConfig.ttl,
        limit: rateLimitConfig.limit,
        storage: new ThrottlerStorageRedisService([
          redisRateLimitConfig.host,
        ], {
          enableReadyCheck: false,
          retryDelayOnClusterDown: 300,
          retryDelayOnFailover: 1000,
          retryDelayOnTryAgain: 3000,
          slotsRefreshTimeout: 10000,
          redisOptions: {
            keyPrefix: '{login-throttler}',
          },
        }),
      };
```

For this package I would suggest maybe separating the cluster part out into it's own class instead of doing them both together as you need cluster options etc which might be application to people in the future.

**Comment by kkoomen** [comment]
@uncodable we can combine your example like so:

NOTE: this is just an idea, haven't tested it yet.

```ts
return {
  ttl: rateLimitConfig.ttl,
  limit: rateLimitConfig.limit,
  storage: new ThrottlerStorageRedisClusterService(
    [redisRateLimitConfig.host],
    {
      enableReadyCheck: false,
      retryDelayOnClusterDown: 300,
      retryDelayOnFailover: 1000,
      retryDelayOnTryAgain: 3000,
      slotsRefreshTimeout: 10000,
      redisOptions: {
        keyPrefix: '{login-throttler}',
      },
    }
  }),
};
```

and then the code will be changed like so:

```diff
diff --git a/src/throttler-storage-redis-cluster.service.ts b/src/throttler-storage-redis-cluster.service.ts
new file mode 100644
index 0000000..6942f77
--- /dev/null
+++ b/src/throttler-storage-redis-cluster.service.ts
@@ -0,0 +1,11 @@
+import { Injectable } from '@nestjs/common';
+import Redis, { Cluster, ClusterNode, ClusterOptions } from 'ioredis';
+import { ThrottlerStorageRedisService } from './throttler-storage-redis.service';
+
+@Injectable()
+export class ThrottlerStorageRedisClusterService extends ThrottlerStorageRedisService {
+  constructor(nodes: ClusterNode[], options?: ClusterOptions, scanCount?: number) {
+    const cluster: Cluster = new Redis.Cluster(nodes, options);
+    super(cluster, scanCount);
+  }
+}
diff --git a/src/throttler-storage-redis.interface.ts b/src/throttler-storage-redis.interface.ts
index 353df80..fd746b7 100644
--- a/src/throttler-storage-redis.interface.ts
+++ b/src/throttler-storage-redis.interface.ts
@@ -1,10 +1,10 @@
-import Redis from 'ioredis';
+import Redis, { Cluster } from 'ioredis';
 
 export interface ThrottlerStorageRedis {
   /**
    * The redis instance.
    */
-  redis: Redis;
+  redis: Redis | Cluster;
 
   /**
    * The amount of items that redis should return for each scan.
diff --git a/src/throttler-storage-redis.service.ts b/src/throttler-storage-redis.service.ts
index a5f554a..42554a7 100644
--- a/src/throttler-storage-redis.service.ts
+++ b/src/throttler-storage-redis.service.ts
@@ -1,20 +1,21 @@
 import { Injectable, OnModuleDestroy } from '@nestjs/common';
-import Redis, { RedisOptions } from 'ioredis';
+import Redis, { Cluster, RedisOptions } from 'ioredis';
 import { ThrottlerStorageRedis } from './throttler-storage-redis.interface';
 
 @Injectable()
 export class ThrottlerStorageRedisService implements ThrottlerStorageRedis, OnModuleDestroy {
-  redis: Redis;
+  redis: Redis | Cluster;
   disconnectRequired?: boolean;
   scanCount: number;
 
   constructor(redis?: Redis, scanCount?: number);
   constructor(options?: RedisOptions, scanCount?: number);
+  constructor(cluster?: Cluster, scanCount?: number);
   constructor(url?: string, scanCount?: number);
-  constructor(redisOrOptions?: Redis | RedisOptions | string, scanCount?: number) {
+  constructor(redisOrOptions?: Redis | RedisOptions | Cluster | string, scanCount?: number) {
     this.scanCount = typeof scanCount === 'undefined' ? 1000 : scanCount;
 
-    if (redisOrOptions instanceof Redis) {
+    if (redisOrOptions instanceof Redis || redisOrOptions instanceof Cluster) {
       this.redis = redisOrOptions;
     } else if (typeof redisOrOptions === 'string') {
       this.redis = new Redis(redisOrOptions as string);
```

**Comment by mpx2m** [comment]
Any planning time to finish the feature of supporting cluster mode?

**Comment by kkoomen** [comment]
@mpx2m Thanks for reminding me, as I'm very busy. I can test the above suggestion I did anytime soon, so I'll try to work on it.

**Comment by kkoomen** [comment]
@mpx2m since I don't have a proper cluster setup, do you? if so, can you test the patch provided above?

**Comment by vjunjoced** [comment]
I would like to know if there is any advance with this integration ?

**Comment by kkoomen** [comment]
@zfanta @mikeulvila @n10000k @mpx2m See #1038 for a possible implementation for the Redis Clusters. Can you all test this and see if this works? Any feedback on the PR #1038 should be commented over there, not here. Thanks.

**Comment by mpx2m** [comment]
@zfanta I implemented the throttler functionality in the service mesh now, not test the code yet.

A docker redis cluster should be enough to test it.

**Comment by kkoomen** [comment]
For those who want to contribute, have a look at #1040 along with the PR #1041

**Comment by kkoomen** [comment]
#1041 fixed a good performance issue, I've merged the code into #1038 but not sure if there's anything more needed in #1038. If anyone could test the changes in #1038 with 2 or more redis clusters, that'd be great.

**Comment by kkoomen** [comment]
I just merged #1038 with proper support for redis clusters along with automated tests. Version [v0.2.1](https://github.com/kkoomen/nestjs-throttler-storage-redis/releases/tag/v0.2.1) and above  contains full support for clusters. If there are any problems, feel free to comment here.

Enjoy.
